### PR TITLE
Add UUID header to mgmt API

### PR DESCRIPTION
**Pull request by shanyp**
## Issue ticket number and link
#3479
## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

### CODE REVIEW

File: libs/utils/src/http/endpoint.rs [hunk 1]
```diff
@@ -63,8 +63,22 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
...
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |mut req| async move {
+        let headers = req.headers_mut();
+        let name = HeaderName::from_str("UUID").unwrap();
+        let request_id = uuid::Uuid::new_v4().to_string();
+        let value = HeaderValue::from_str(&request_id).unwrap();
+        headers.insert(name, value);
+        tracing::info!("{} {} {}", req.method(), req.uri().path(), request_id);
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:74]: Should it be debug or something else to avoid increase in log volume?

See https://github.com/neondatabase/neon/pull/3681

File: libs/utils/src/http/endpoint.rs [hunk 2]
```diff
@@ -63,8 +63,22 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
     Ok(response)
 }
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |mut req| async move {
+        let headers = req.headers_mut();
+        let name = HeaderName::from_str("UUID").unwrap();
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:70]: NIT: lets use expect instead of unwrap

File: libs/utils/src/http/endpoint.rs [hunk 3]
```diff
@@ -11,7 +11,7 @@ use routerify::ext::RequestExt;
 use routerify::RequestInfo;
 use routerify::{Middleware, Router, RouterBuilder, RouterService};
 use tokio::task::JoinError;
-use tracing;
+use tracing::{self};
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:14]: Is this change required?

**Review by LizardWizzard — COMMENTED** [review]
The issue additionally mentions creating a span for each handler and putting the request id into span variables

File: libs/utils/src/http/endpoint.rs [hunk 4]
```diff
@@ -63,8 +63,26 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
...
 }
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |mut req| async move {
+        let headers = req.headers_mut();
+        let name = HeaderName::from_str("UUID").expect("created header name");
+        let request_id = uuid::Uuid::new_v4().to_string();
+        let value = HeaderValue::from_str(&request_id).unwrap();
+        headers.insert(name, value);
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:73]: Also I think we need to insert header only if it is missing. It will allow client to specify the id and it can be used to have end to end request id

File: libs/utils/src/http/endpoint.rs [hunk 5]
```diff
@@ -63,8 +63,37 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
...
+
+        let headers = req.headers_mut();
+        let request_id = match headers.get(&x_request_id) {
+            Some(request_id) => request_id
+                .to_str()
+                .expect("extract request id value")
+                .to_owned(),
+            None => {
+                let request_id = uuid::Uuid::new_v4().to_string();
+                let value = HeaderValue::from_str(&request_id).unwrap();
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:79]: Lets have expect here as well

File: libs/utils/src/http/endpoint.rs [hunk 6]
```diff
@@ -63,8 +63,37 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
     Ok(response)
 }
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |mut req| async move {
+        let x_request_id = HeaderName::from_static("x-request-id");
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:69]: Is it the same as in console?

**Review by LizardWizzard — APPROVED** [review]
LGTM, minor nits

File: libs/utils/src/http/endpoint.rs (hunk 6 above)
Comment (shanyp) [reply libs/utils/src/http/endpoint.rs:69]: yes : https://github.com/neondatabase/cloud/blob/189ed5819bf6c9c35038a3462c7604585b349e28/internal/resttooling/requestid/requestid.go#L16

**Review by LizardWizzard — APPROVED** [review]

File: libs/utils/src/http/endpoint.rs [hunk 7]
```diff
@@ -27,14 +28,38 @@ static SERVE_METRICS_COUNT: Lazy<IntCounter> = Lazy::new(|| {
     .expect("failed to define a metric")
 });
 
+static X_REQUEST_ID_HEADER_STR: &str = "x-request-id";
+
+static X_REQUEST_ID_HEADER: HeaderName = HeaderName::from_static(X_REQUEST_ID_HEADER_STR);
+#[derive(Debug, Default, Clone)]
+struct RequestId {
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:35]: 1. Box<str> should be a slightly better option (a bit less size since its immutable)
2. There are several options how to declare that. You can use newtype: 

```rust
struct RequestId(Box<str>)
```
Though .id will become .0

Completely optional, up to you

File: libs/utils/src/http/endpoint.rs [hunk 8]
```diff
@@ -27,14 +28,38 @@ static SERVE_METRICS_COUNT: Lazy<IntCounter> = Lazy::new(|| {
...
+    id: String,
+}
+
 async fn logger(res: Response<Body>, info: RequestInfo) -> Result<Response<Body>, ApiError> {
     // cannot factor out the Level to avoid the repetition
     // because tracing can only work with const Level
     // which is not the case here
+    let request_id_val = info.context::<RequestId>();
+
+    let request_id = request_id_val.unwrap_or_default().id;
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:45]: NIT: I would rather squash these together without temp `request_id_val`

File: libs/utils/src/http/endpoint.rs [hunk 9]
```diff
@@ -63,9 +88,53 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
...
+                Cow::Borrowed(request_id.to_str().expect("extract request id value"))
+            }
+            None => {
+                let request_id = uuid::Uuid::new_v4();
+                Cow::Owned(request_id.to_string())
+            }
+        };
+
+        req.set_context(RequestId {
+            id: request_id.to_string(),
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:105]: Here we copy the value anyway, so Cow is no longer saves us an allocation. Sorry for going back and forth with it

**Review by LizardWizzard — APPROVED** [review]
LGTM

Left a bunch of NITs, following them is up to you

File: libs/utils/Cargo.toml [hunk 10]
```diff
@@ -13,6 +13,7 @@ bincode.workspace = true
 bytes.workspace = true
 heapless.workspace = true
 hyper = { workspace = true, features = ["full"] }
+futures = { version = "0.3" }
```
Comment (LizardWizzard) [inline libs/utils/Cargo.toml:16]: We have futures in workspace, so should be ok with `futures.workspace = true`

File: libs/utils/src/http/endpoint.rs [hunk 11]
```diff
@@ -300,3 +299,55 @@ where
...
+        let remote_addr = SocketAddr::new(IpAddr::from_str("127.0.0.1").unwrap(), 80);
+        let mut service = builder.build(remote_addr);
+        if let Err(e) = poll_fn(|ctx| service.poll_ready(ctx)).await {
+            panic!("request service is not ready: {:?}", e);
+        }
+
+        let mut req: Request<Body> = Request::default();
+        req.headers_mut().append(
+            &X_REQUEST_ID_HEADER,
+            HeaderValue::from_str(X_REQUEST_ID_HEADER_STR).unwrap(),
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:322]: NIT: A bit confusing on the first glance, can be `from_str("42")` or something

File: libs/utils/src/http/endpoint.rs [hunk 12]
```diff
@@ -300,3 +299,55 @@ where
...
+            panic!("request service is not ready: {:?}", e);
+        }
+
+        let req: Request<Body> = Request::default();
+        let resp: Response<hyper::body::Body> = service.call(req).await.unwrap();
+
+        let header_val = resp.headers().get(&X_REQUEST_ID_HEADER);
+
+        if let Some(_header) = header_val {
+            assert!(true, "response header should be empty");
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:350]: Dont get it. I thought that it should always be there and non empty (our own uuid if client didnt provide one)

File: libs/utils/src/http/endpoint.rs [hunk 13]
```diff
@@ -63,9 +87,53 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
...
+        }
+        Ok(req)
+    })
+}
+
+async fn add_request_id_header_to_response(
+    mut res: Response<Body>,
+    req_info: RequestInfo,
+) -> Result<Response<Body>, ApiError> {
+    if let Some(request_id) = req_info.context::<RequestId>() {
```
Comment (LizardWizzard) [inline libs/utils/src/http/endpoint.rs:120]: Thinking about defensive checks here. Not sure we should add them, maybe debug_assert or something. Should be fine as is too

File: libs/utils/src/http/endpoint.rs [hunk 14]
```diff
@@ -27,14 +28,38 @@ static SERVE_METRICS_COUNT: Lazy<IntCounter> = Lazy::new(|| {
     .expect("failed to define a metric")
 });
 
+static X_REQUEST_ID_HEADER_STR: &str = "x-request-id";
```
[the remark this answers is not in the document]
Comment (shanyp) [reply libs/utils/src/http/endpoint.rs:30]: why? it is used in other places in the code

File: libs/utils/src/http/endpoint.rs (hunk 12 above)
Comment (shanyp) [reply libs/utils/src/http/endpoint.rs:350]: yep

File: libs/utils/src/http/endpoint.rs (hunk 13 above)
Comment (shanyp) [reply libs/utils/src/http/endpoint.rs:115]: we don't want to fail if its missing
